### ISSUE TITLE
loss calculation for PaliGemmaForConditionalGeneration potentially not cast to correct device

### ISSUE THREAD

**Issue by grahamannett** [huggingface/transformers#30997]
### System Info

```
Name: transformers
Version: 4.42.0.dev0
```

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Use `PaliGemmaForConditionalGeneration.forward` with device_map='auto' and labels for 4+ devices.

### Expected behavior

I believe which device logits/labels/attention_mask will be on for `PaliGemmaForConditionalGeneration` can potentially cause an error for others (and is for me as well by default as well, but I have patched it) when using `device_map='auto'`.

https://github.com/huggingface/transformers/blob/main/src/transformers/models/paligemma/modeling_paligemma.py#L487

```
                shift_logits = shift_logits[shift_attention_mask.to(logits.device) != 0].contiguous()
                shift_labels = shift_labels[shift_attention_mask.to(logits.device) != 0].contiguous()
```

labels and attention_mask should probably be on the lowest index device.  logits likely to be a later device if the model is sharded with something like `device_map='auto'`.  The shift_logits line is right since the shift_logits come from the logits but the shift_labels line can cause issues since shift_labels should likely be moved to logits device.

**Comment by molbap** [comment]
Hi @grahamannett , thanks for the catch - will open a PR to fix!

**Comment by ArthurZucker** [comment]
@grahamannett can you share the traceback as well? To identify where this failed
